source/common/router/config_utility.cc

```diff
@@ -38,21 +38,33 @@ maybeCreateStringMatcher(const envoy::config::route::v3::QueryParameterMatcher&
 ConfigUtility::QueryParameterMatcher::QueryParameterMatcher(
     const envoy::config::route::v3::QueryParameterMatcher& config,
     Server::Configuration::CommonFactoryContext& context)
-    : name_(config.name()), matcher_(maybeCreateStringMatcher(config, context)) {}
+    : name_(config.name()),
+      present_match_(config.has_present_match() ? absl::make_optional(config.present_match())
+                                                : absl::nullopt),
+      matcher_(maybeCreateStringMatcher(config, context)) {}
 
 bool ConfigUtility::QueryParameterMatcher::matches(
     const Http::Utility::QueryParamsMulti& request_query_params) const {
   // This preserves the legacy behavior of ignoring all but the first value for a given key
   auto data = request_query_params.getFirstValue(name_);
+
+  // If we're doing a present_match, return whether the parameter exists and matches the expected
```

Comment:
Yeah this looks more correct to me.  As we can't tell if folks are depending on the long-standing buggy behavior, please runtime guard
/wait

Reply:
Done